File: sdk/cpp/examples/example_server.cpp
// examples/example_server.cpp — a type-theoretic RCP server.
// Build: c++ -std=c++23 -Iinclude examples/example_server.cpp -o example_server
// Run:   ./example_server            (stdio)   |   ./example_server --http 8000
#include "rcp.hpp"

#include <cmath>
#include <string>
#include <vector>

using namespace rcp;

namespace {
constexpr std::size_t DIM = 384;

std::vector<float> embed_one(const std::string& text) {
    std::vector<float> v(DIM, 0.0f);
    std::string tok;
    auto flush = [&] {
        if (tok.empty()) return;
        std::uint64_t h = 1469598103934665603ull;
        for (char c : tok) { h ^= (unsigned char)std::tolower(c); h *= 1099511628211ull; }
        v[h % DIM] += 1.0f; tok.clear();
    };
    for (char c : text) { if (std::isspace((unsigned char)c)) flush(); else tok.push_back(c); }
    flush();
    float n = 0; for (float x : v) n += x * x; n = n > 0 ? 1.0f / std::sqrt(n) : 0;
    for (float& x : v) x *= n;
    return v;
}
float cosine(const std::vector<float>& a, const std::vector<float>& b) {
    float s = 0; for (std::size_t i = 0; i < a.size(); ++i) s += a[i] * b[i]; return s;
}
const std::vector<std::pair<std::string, std::string>> DOCS = {
    {"d1", "The Eiffel Tower is a wrought-iron lattice tower in Paris, France."},
    {"d2", "Photosynthesis converts light energy into chemical energy in plants."},
    {"d3", "The Great Wall of China stretches thousands of kilometres."},
};
} // namespace

// A handler models the Handler concept: info + capabilities + whichever method
// hooks it implements. Advertising a capability whose hook you didn't write
// would just yield CapabilityMissing at dispatch — no null-pointer surprises.
struct ExampleEngine {
    PeerInfo info() const { return {"rcp-cpp-example", "1.0.0"}; }

    Capabilities capabilities() const {
        Capabilities c;
        c.with_embed(Dimension{DIM}, "toy-hash-v1")
         .with_retrieve(100, {"dense", "sparse", "hybrid"})
         .with_graph({"local", "global"})
         .with_session()
         .with_feedback()
         .with_memory({"global"}, true);
        c.citations = true;
        return c;
    }

    Result<Json> embed(const Json& p) {
        // Accept `inputs` (preferred) or the legacy `texts`; each element is a
        // bare string or a Content block (spec §4.8 / §7.3).
        const Json& in = p.contains("inputs") ? p["inputs"] : p.value("texts", Json::array());
        Json vecs = Json::array();
        for (auto& t : in) {
            std::string s = t.is_string() ? t.get<std::string>() : t.value("text", std::string{});
            vecs.push_back(embed_one(s));
        }
        return Json{{"vectors", std::move(vecs)}};
    }

    Result<Json> retrieve(const Json& p) {
        std::string q = p.value("query", std::string{});
        std::size_t k = p.value("k", std::size_t{10});
        auto qv = embed_one(q);
        std::vector<std::pair<float, std::size_t>> scored;
        for (std::size_t i = 0; i < DOCS.size(); ++i) scored.push_back({cosine(qv, embed_one(DOCS[i].second)), i});
        std::sort(scored.begin(), scored.end(), [](auto& a, auto& b) { return a.first > b.first; });
        Json hits = Json::array();
        for (std::size_t r = 0; r < scored.size() && r < k; ++r) {
            auto& [s, i] = scored[r];
            hits.push_back(Json{{"id", DOCS[i].first}, {"score", s}, {"text", DOCS[i].second},
                                {"confidence", s < 0 ? 0.0 : (s > 1 ? 1.0 : s)},  // normalised [0,1]
                                {"unit", p.value("unit", std::string{"chunk"})},
                                {"citation", {{"source", DOCS[i].first}}},
                                {"trust", {{"level", "trusted"}}}});
        }
        return Json{{"hits", std::move(hits)},
                    {"usage", {{"mode", p.value("mode", std::string{"hybrid"})},
                               {"candidates", DOCS.size()},
                               {"reranked", 0}}}};
    }

    Result<Json> graph(const Json& p) {
        if (p.value("op", std::string{}) == "global")
            return Json{{"summary", "a global community summary"}, {"communities", Json::array()}};
        return retrieve(p);
    }

    // Client→server signals (spec §7.16): a real engine would log or online-update;
    // the example just counts what it accepted.
    Result<Json> feedback(const Json& p) {
        std::size_t n = p.contains("signals") && p["signals"].is_array() ? p["signals"].size() : 0;
        return Json{{"accepted", n}};
    }

    // Global/session memory (spec §7.17): a real engine would summarise the corpus;
    // the example returns a stub handle and a couple of clues.
    Result<Json> memory_build(const Json&) {
        return Json{{"memoryId", "mem-example"}, {"tokens", 0}};
    }
    Result<Json> memory_recall(const Json& p) {
        return Json{{"clues", Json::array({
            Json{{"query", p.value("query", std::string{})}, {"weight", 1.0}},
            Json{{"seedIds", Json::array({"d1"})}, {"weight", 0.5}}})}};
    }
};

int main(int argc, char** argv) {
    Server srv{ExampleEngine{}};
    if (argc >= 3 && std::string(argv[1]) == "--http") {
        auto r = srv.serve_http(static_cast<std::uint16_t>(std::atoi(argv[2])));
        return r ? 0 : 1;
    }
    srv.serve_stdio();
    return 0;
}

File: examples/example_server.py
#!/usr/bin/env python3
"""Example RCP/1 server — a complete, dependency-free HYBRID retrieval engine.

Run as a subprocess (stdio, the default):
    python3 examples/example_server.py
Or over HTTP:
    python3 examples/example_server.py --http 8000

This is a faithful miniature of a production RAG pipeline (spec §3), built with
nothing but the standard library so you can read the whole thing:

    query ─▶ dense recall  ┐
            sparse recall  ├─▶ RRF fusion ─▶ rerank ─▶ top-k hits + citations
                           ┘   (§16.3)      (§7.6)      (the funnel, §3.3)

It advertises **embed + sparseEmbed + rerank + retrieve + graph**, supports
`mode: dense|sparse|hybrid`, honours the `candidateK ≥ rerank.topN ≥ k` funnel,
returns per-stage `scores` and `usage`, and certifies at conformance **L2**.
Swap the toy embedders for sentence-transformers / SPLADE / FAISS / a
cross-encoder and a real cross-encoder — the wire contract never changes.
"""
import hashlib
import math
import os
import re
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "sdk", "python"))

import rcp  # noqa: E402
from rcp import rrf_fuse  # the SDK's reference fusion (spec §16.3)  # noqa: E402

DIM = 384
DOCS = [
    ("d1", "The Eiffel Tower is a wrought-iron lattice tower on the Champ de Mars in Paris, France.", "https://en.wikipedia.org/wiki/Eiffel_Tower"),
    ("d2", "Photosynthesis converts light energy into chemical energy stored in glucose within plant chloroplasts.", "https://en.wikipedia.org/wiki/Photosynthesis"),
    ("d3", "The Great Wall of China stretches thousands of kilometres across the country's northern borders.", "https://en.wikipedia.org/wiki/Great_Wall_of_China"),
    ("d4", "Paris is the capital of France and sits on the river Seine in the north of the country.", "https://en.wikipedia.org/wiki/Paris"),
    ("d5", "Iron is a chemical element; wrought iron is a tough, malleable form used in historic construction.", "https://en.wikipedia.org/wiki/Wrought_iron"),
]

_WORD = re.compile(r"[a-z0-9]+")


def _tokens(text):
    return _WORD.findall(text.lower())


# ── dense embedder: hashed bag-of-words, L2-normalised (stands in for a model) ──
def embed_dense(text):
    v = [0.0] * DIM
    for tok in _tokens(text):
        h = int(hashlib.blake2b(tok.encode(), digest_size=8).hexdigest(), 16)
        v[h % DIM] += 1.0
    n = math.sqrt(sum(x * x for x in v)) or 1.0
    return [x / n for x in v]


# ── sparse embedder: term → weight, SPLADE-style (indices are hashed vocab ids) ─
def embed_sparse(text):
    weights = {}
    for tok in _tokens(text):
        idx = int(hashlib.blake2b(tok.encode(), digest_size=4).hexdigest(), 16)
        weights[idx] = weights.get(idx, 0.0) + 1.0
    items = sorted(weights.items())
    return {"indices": [i for i, _ in items], "values": [w for _, w in items]}


_DENSE = {uri: embed_dense(text) for uri, text, _ in DOCS}
_SPARSE = {uri: embed_sparse(text) for uri, text, _ in DOCS}
_TEXT = {uri: text for uri, text, _ in DOCS}
_URI = {uri: src for uri, _, src in DOCS}


def _cosine(a, b):
    return sum(x * y for x, y in zip(a, b))


def _sparse_dot(q, d):
    dv = dict(zip(d["indices"], d["values"]))
    return sum(w * dv.get(i, 0.0) for i, w in zip(q["indices"], q["values"]))


def _dense_ranked(query, n):
    q = embed_dense(query)
    scored = sorted(((_cosine(q, _DENSE[u]), u) for u in _DENSE), reverse=True)
    return [{"id": u, "score": s} for s, u in scored[:n]]


def _sparse_ranked(query, n):
    q = embed_sparse(query)
    scored = sorted(((_sparse_dot(q, _SPARSE[u]), u) for u in _SPARSE), reverse=True)
    return [{"id": u, "score": s} for s, u in scored[:n] if s > 0] or \
           [{"id": u, "score": s} for s, u in scored[:n]]


def _rerank_score(query, text):
    # A toy cross-encoder: lexical overlap × length prior. Stands in for a real
    # query-document interaction model (§7.6). Deterministic and in [0,1].
    q, d = set(_tokens(query)), set(_tokens(text))
    if not q:
        return 0.0
    return len(q & d) / len(q)


def _retrieve(query, k, candidate_k, top_n, mode):
    """Run the funnel: recall (mode) → fuse → rerank → top-k (spec §3, §7.7)."""
    # 1. Recall — wide and cheap. `candidateK` candidates per enabled retriever.
    lists = {}
    if mode in ("dense", "hybrid"):
        lists["dense"] = _dense_ranked(query, candidate_k)
    if mode in ("sparse", "hybrid"):
        lists["sparse"] = _sparse_ranked(query, candidate_k)

    # 2. Fuse — rank-based, so dense cosine and sparse dot never get compared
    #    directly. Single-retriever modes pass through fusion unchanged.
    fused = rrf_fuse(lists, k=None)
    per_stage = {u["id"]: dict(u.get("meta", {})) for u in fused}

    # 3. Rerank — precise and narrow. Rescore the top `top_n` candidates.
    head = fused[:top_n]
    for h in head:
        h["_rr"] = _rerank_score(query, _TEXT[h["id"]])
    head.sort(key=lambda h: h["_rr"], reverse=True)

    # 4. Pack + cite — return the best `k`, richest body first.
    hits = []
    for h in head[:k]:
        uri = h["id"]
        stage = per_stage.get(uri, {})
        hits.append({
            "id": uri,
            "score": round(h["_rr"], 4),
            "text": _TEXT[uri],
            "scores": {kk: round(vv, 4) for kk, vv in stage.items()
                       if isinstance(vv, (int, float))} | {"rerank": round(h["_rr"], 4)},
            "citation": {"source": uri, "uri": _URI[uri], "title": _TEXT[uri][:40] + "…"},
        })
    usage = {"candidates": len(fused), "reranked": len(head), "mode": mode,
             "retrievers": list(lists.keys())}
    return {"hits": hits, "usage": usage, "indexVersion": "demo-1"}


def build():
    s = rcp.Server()
    s.set_info("rcp-example", "1.0.0")
    # Declare L2 honestly — the conformance harness will hold us to it.
    s.set_conformance("L2")
    s.advertise(rcp.Capability.Embed, {"dimension": DIM, "modalities": ["text"],
                                       "encodings": ["json", "f32-base64"]})
    s.advertise(rcp.Capability.SparseEmbed, {"identity": "demo-splade"})
    s.advertise(rcp.Capability.Rerank, {"methods": ["cross-encoder"], "maxCandidates": 100})
    s.advertise(rcp.Capability.Retrieve, {"maxK": 100, "modes": ["dense", "sparse", "hybrid"],
                                          "citations": True, "defaultCandidateK": 50})
    s.advertise(rcp.Capability.Graph, {"ops": ["local", "global"]})

    @s.on(rcp.Method.EMBED)
    def _embed(params):
        items = params.get("inputs") or params.get("texts", [])
        texts = [t if isinstance(t, str) else t.get("text", "") for t in items]
        vecs = [embed_dense(t) for t in texts]
        encoding = params.get("encoding", "json")
        payload, meta = rcp.encode_vectors(vecs, encoding)
        return {"vectors": payload, **meta}

    @s.on(rcp.Method.EMBED_SPARSE)
    def _embed_sparse(params):
        items = params.get("texts") or params.get("inputs", [])
        return {"sparse": [embed_sparse(t if isinstance(t, str) else t.get("text", ""))
                           for t in items]}

    @s.on(rcp.Method.RERANK)
    def _rerank(params):
        query = params.get("query", "")
        docs = params.get("documents", [])
        top_n = int(params.get("topN", len(docs)))
        scored = sorted(
            ({"index": i, "score": round(_rerank_score(query, d if isinstance(d, str) else d.get("text", "")), 4)}
             for i, d in enumerate(docs)),
            key=lambda r: r["score"], reverse=True)
        return {"results": scored[:top_n]}

    @s.on(rcp.Method.RETRIEVE)
    def _retrieve_handler(params):
        k = int(params.get("k", 5))
        mode = params.get("mode", "hybrid")
        # Funnel defaults: candidateK ≥ topN ≥ k, all server-chosen when absent.
        candidate_k = int(params.get("candidateK", max(k, min(50, len(DOCS)))))
        rr = params.get("rerank") or {}
        top_n = int(rr.get("topN", max(k, min(candidate_k, 20))))
        if mode not in ("dense", "sparse", "hybrid"):
            raise rcp.RcpError(rcp.Errc.OPTION_UNSUPPORTED, f"mode '{mode}' not supported",
                               {"option": "mode"})
        return _retrieve(params.get("query", ""), k, candidate_k, top_n, mode)

    @s.on(rcp.Method.GRAPH)
    def _graph(params):
        op = params.get("op", "local")
        if op == "local":
            r = _retrieve(params.get("query", ""), int(params.get("k", 5)),
                          candidate_k=len(DOCS), top_n=len(DOCS), mode="hybrid")
            return {"hits": r["hits"]}
        if op == "global":
            return {"summary": "One community: European landmarks and plant biology.",
                    "communities": [{"id": "c0", "summary": "landmarks + biology", "score": 1.0}]}
        return {"op": op, "note": "unimplemented"}

    return s


if __name__ == "__main__":
    srv = build()
    if len(sys.argv) >= 3 and sys.argv[1] == "--http":
        srv.serve_http(int(sys.argv[2]))
    else:
        srv.serve_stdio()
